File: Local/Matlab/mfile/plot_drm_visit_times.m
function plot_drm_visit_times(dest_tmpl, mode, t_info, t_visit_time)
%plot_drm_visit_times	plot detection/charaterization visits in a drm-set
% 
% plot_drm_visit_times(dest_tmpl, mode, t_info, t_visit_time)
% * Time-series plots of visits for the purpose of detection or
% characterization - (#visits) plotted vs. mission elapsed time.
% * Note: This plot family is similar to yield, but is *not* yield:
%   t_yield_time -- yield (#planets) versus mission elapsed time. 
%   t_visit_time -- star-visits versus mission elapsed time.
% So, "visits" are counted whether they result in success or not, and 
% visits do not count planets, they count periods-of-integration.
%
% Inputs:
%   string dest_tmpl
%   string mode
%   table t_info
%   table t_visit_time
% 
% Outputs:
%   (to disk)
% 
% See Also:  

% 
% Error checking
% 
% if all(nargin  ~= [3]), error ('Bad input arg number'); end
% if all(nargout ~= [0 1]), error ('Bad output arg number'); end  
% allow passing an empty metadata table as []
if ismatrix(t_info) && isempty(t_info),
    t_info = cell2table(cell(0));
end;

% skip, unless mode.op contains our name or a *
if length(strfind(mode.op, '*')) == 0 && ...
   length(strfind(mode.op, 'visit_time')) == 0,
    fprintf('Visit time plots: skipping, as directed.\n');
    return;
end

% file extensions to write
%ExtList = {'png', 'pdf'};
ExtList = {'png'};


%
% Utilities
% 

% Inner function:
%   Set up plot/axis styles, title, axis labels.
%   Uses global mfile variables, such as t_info
%   Requires title, and ylabel text, and legend text (cell)
    function style_visit_plot(title1, ytext, legtext)
    
    % errorbars can force Y range < 0, so clip them at 0
    set(gca, 'YLim', max(0, get(gca, 'YLim'))); 
    % format the title
    title2 = plot_make_title(t_info);

    % title: prevent special interpretation of _
    title({title2, title1}, 'Interpreter', 'none');
    % axis labels
    xlabel('Time [day]', 'FontWeight', 'bold')
    ylabel(ytext, 'FontWeight', 'bold')
    legend(legtext, 'Interpreter', 'none', 'Location', 'northwest');
    set(gca,'TitleFontSizeMultiplier', 1.1)
    set(gca, 'FontSize', 13);
    grid on;
    end % inner function


% Inner function:
%   write the current figure to various files
%   uses the dest_tmpl mfile variable
    function write_plots(dest_name)

    % write the plot out
    if ~isempty(dest_tmpl),
        for ext = ExtList,
            fn_gfx = sprintf(dest_tmpl, dest_name, ext{1});
            fprintf(1, '\tExporting to %s\n', fn_gfx);
            export_fig('-transparent', '-m2', fn_gfx);
        end;
    end
    end

%
% Computation
% 

% offsets on various error-bars in the plot 
%  (units of days with one complete sample per ~30 days)
t_offsets = [0 10 5];
% sample times
tsamp = t_visit_time{:, 'h_det_time_lo'};


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Plot roster

% Needed for each item here: 
%  {fieldnames, plot title, filename}

FigRoster = { ...
    { ...
        {'h_visit_det_visit', 'h_visit_det_uniq', 'h_visit_det_revi'}, ...
        'Detection-Mode Target Visits', 'visit-time-det' ... 
    }, ...
    { ...
        {'h_visit_char_visit', 'h_visit_char_uniq', 'h_visit_char_revi'}, ...
        'Characterization-Mode Target Visits', 'visit-time-char' ... 
    }, ...
             };
    
% iterate over all figures in the roster

for fig_num = 1:length(FigRoster),
    FigInfo = FigRoster{fig_num};
    [names, dtxt, fname] = FigInfo{:};

    % lines in the plot
    N_plot = length(names);

    % skip unless the required fieldnames are present
    % (e.g., no char summaries are in t_visit_time for det-only missions)
    skipping = false;
    for n = 1:N_plot,
        f_mean = sprintf('%s_%s', names{n}, 'mean');
        if ~any(strcmp(fieldnames(t_visit_time), f_mean)),
            skipping = true;
        end
    end;
    if skipping,
        fprintf(sprintf('\tSkipping %s plot (missing field)\n', fname));
        continue;
    end

    %% A: Monthly plot
    clf; set(gcf, 'Position', [100 100 850 500]);
   
    names_legend = {sprintf('All %s', dtxt), ...
                    sprintf('First-time %s', dtxt), ...
                    sprintf('Return %s', dtxt) };

    % put each of the above timeseries on one plot
    for n = 1:N_plot,
        f = names{n};
        f_mean = sprintf('%s_%s', f, 'mean');
        f_std  = sprintf('%s_%s', f, 'std');
        h_eb = errorbar(tsamp+t_offsets(n), t_visit_time{:,f_mean}, t_visit_time{:,f_std});
        % style the plot
        set(h_eb, 'LineWidth', 1); % NB: skinny
        hold on;
    end;

    style_visit_plot(...
        sprintf('Monthly %s vs. Mission Time', dtxt), ...
        sprintf('%s [count/month]', dtxt), ...
        names_legend);
    write_plots([fname '-month']);


    %% B: Cumulative plot
    % Note: we derive the error bars here

    clf; set(gcf, 'Position', [100 100 850 500]);

    % put each of the above timeseries on one plot
    for n = 1:N_plot,
        f = names{n};
        f_mean = sprintf('%s_%s', f, 'mean');
        f_std  = sprintf('%s_%s', f, 'std');
        % cumulative mean, plus cumulative std error (adding in quadrature,
        % i.e., sqrt-sum-of-squares)
        h_eb = errorbar(tsamp+t_offsets(n), ...
                        cumsum(t_visit_time{:,f_mean}), sqrt(cumsum(t_visit_time{:,f_std}.^2)));
        % style the plot
        set(h_eb, 'LineWidth', 2);
        hold on;
    end;

    style_visit_plot(...
        sprintf('Cumulative %s vs. Mission Time', dtxt), ...
        sprintf('%s [count]', dtxt), ...
        names_legend);
    write_plots([fname '-cume']);

end % of figure for-loop

return
end % of main function
    

File: util/plot_drm_gallery/plot_drm_visit_times.py
#!/usr/bin/env python
"""
Plot detection/characterization visits in a drm-set

Time-series plots of visits for the purpose of detection or
characterization - (#visits) plotted vs. mission elapsed time.

Note: This plot family is similar to yield, but is *not* yield:
  t_yield_time -- yield (#planets) versus mission elapsed time. 
  t_visit_time -- star-visits versus mission elapsed time.
So, "visits" are counted whether they result in success or not, and 
visits do not count planets, they count periods-of-integration.
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import argparse
import sys
import os
# uniform plot appearance
try:
    from . import common_style as cs
except ImportError:
    import common_style as cs

# Program name for error messages
PROGNAME = os.path.basename(sys.argv[0])


def plot_drm_visit_times(reduce_info, plot_data, dest_tmpl, mode):
    """
    Plot detection/characterization visits in a drm-set

    Time-series plots of visits for the purpose of detection or
    characterization - (#visits) plotted vs. mission elapsed time.

    Parameters
    ----------
    reduce_info : dict
        Metadata dict from reduce-info.csv
    plot_data : list of DataFrame
        Pre-loaded CSV data [visit-time]
    dest_tmpl : str
        Template string for output file paths with two %s placeholders
    mode : dict
        Dictionary with 'op' key containing operation mode string

    Outputs
    -------
    Saves plots to disk with names:
        visit-time-det-cume.png
        visit-time-char-cume.png
    """

    # Unpack CSV data
    t_visit_time, = plot_data
    
    # Allow skipping this way
    if '0' in mode.get('op', ''):
        print('Visit time plots: skipping, as directed.')
        return []
    
    # File extensions to write
    ext_list = ['png']

    # Track output files
    tracker = cs.PlotTracker()
    tracker.set_ext_list(ext_list)

    # Inner function: Set up plot/axis styles, title, axis labels
    def style_visit_plot(ax, title1, ytext, legtext):
        """Style the visit plot with title, labels, and legend"""
        # Clip Y range at 0
        ylim = ax.get_ylim()
        ax.set_ylim(max(0, ylim[0]), ylim[1])
        
        # Format the title
        title2 = cs.plot_make_title(reduce_info)
        
        # Set title (preventing special interpretation of _) with bold
        ax.set_title(f'{title2}\n{title1}', fontsize=11*1.1, fontweight='bold')
        
        # Axis labels
        ax.set_xlabel('Time [day]', fontweight='bold')
        ax.set_ylabel(ytext, fontweight='bold')
        
        # Legend
        ax.legend(legtext, loc='upper left')
        ax.tick_params(labelsize=13)
        ax.grid(True)
    
    # Inner function: write the current figure to files
    def write_plots(fig, dest_name):
        """Write the current figure to various files"""
        tracker.write_plots(fig, dest_name, dest_tmpl, verbose=mode['verbose'])

    # Offsets on various error-bars in the plot
    # (units of days with one complete sample per ~30 days)
    t_offsets = [0, 10, 5]

    # Sample times
    try:
        tsamp = t_visit_time['h_det_time_lo'].values
    except KeyError as e:
        print(f"{PROGNAME}: Fatal: Missing required column in visit_time file: {e}", 
              file=sys.stderr)
        sys.exit(1)
    
    # Manual line color order
    # line_colors = ['tab:blue', 'tab:red', 'tab:orange']
    line_colors = ['tab:blue', 'tab:green', 'tab:orange']

    # Uniform error bar properties
    ebar_props = {"marker": '.',
                  "linewidth": 1.7,
                  "errorevery": 1,
                  "elinewidth": 1,
                  "capsize": 1.5}
    
    ####################################################################
    # Plot roster
    ####################################################################
    
    # Needed for each item here: 
    # {fieldnames, plot title, filename}
    
    fig_roster = [
        {
            'names': ['h_visit_det_visit', 'h_visit_det_uniq', 'h_visit_det_revi'],
            'title': 'Detection-Mode Target Visits',
            'fname': 'visit-time-det'
        },
        {
            'names': ['h_visit_char_visit', 'h_visit_char_uniq', 'h_visit_char_revi'],
            'title': 'Characterization-Mode Target Visits',
            'fname': 'visit-time-char'
        }
    ]
    
    # Do not make incremental/monthly plots
    do_incremental_plot = False
    
    # Iterate over all figures in the roster
    for fig_info in fig_roster:
        names = fig_info['names']
        dtxt = fig_info['title']
        fname = fig_info['fname']
        
        n_plot = len(names)
        
        # Skip unless the required fieldnames are present
        # (e.g., no char summaries are in t_visit_time for det-only missions)
        skipping = False
        for name in names:
            f_mean = f'{name}_mean'
            if f_mean not in t_visit_time.columns:
                skipping = True
                break
        
        if skipping:
            print(f'\tSkipping {fname} plot (missing field)')
            continue
        
        # Legend names
        names_legend = [
            f'All {dtxt}',
            f'First-time {dtxt}',
            f'Return {dtxt}'
        ]
        
        # A: Monthly plot (currently disabled)
        if do_incremental_plot:
            fig, ax = plt.subplots(figsize=(8.5, 5))
            
            # Plot each timeseries
            for n, name in enumerate(names):
                f_mean = f'{name}_mean'
                f_std = f'{name}_std'
                
                ax.errorbar(tsamp + t_offsets[n], 
                            t_visit_time[f_mean].values,
                            yerr=t_visit_time[f_std].values,
                            color=line_colors[n],
                            **ebar_props)  # NB: skinny
            
            style_visit_plot(
                ax,
                f'Monthly {dtxt} vs. Mission Time',
                f'{dtxt} [count/month]',
                names_legend
            )
            write_plots(fig, f'{fname}-month')
            plt.close(fig)
        
        # B: Cumulative plot
        # Note: we derive the error bars here
        
        fig, ax = plt.subplots(figsize=(8.5, 5))
        
        # Plot each timeseries with cumulative values
        for n, name in enumerate(names):
            f_mean = f'{name}_mean'
            f_std = f'{name}_std'
            
            # Cumulative mean, plus cumulative std error (adding in quadrature,
            # i.e., sqrt-sum-of-squares)
            cume_mean = np.cumsum(t_visit_time[f_mean].values)
            cume_std = np.sqrt(np.cumsum(t_visit_time[f_std].values ** 2))
            
            ax.errorbar(tsamp + t_offsets[n],
                        cume_mean,
                        yerr=cume_std,
                        color=line_colors[n],
                        **ebar_props)
        
        style_visit_plot(
            ax,
            f'Cumulative {dtxt} vs. Mission Time',
            f'{dtxt} [count]',
            names_legend
        )
        write_plots(fig, f'{fname}-cume')
        plt.close(fig)

    return tracker.get_files()


def main():
    """
    Command-line interface for plot_drm_visit_times
    """
    parser = argparse.ArgumentParser(
        description='Plot detection/characterization visits in a DRM-set',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Example usage:
    python plot_drm_visit_times.py "data/%s.%s" "output/det-%s.%s"
    
The first argument is the source template with two %%s placeholders that will be
filled with ("info", "csv") and ("visit_time", "csv") to locate input files.

The second argument is the destination template with two %%s placeholders for
the plot name and file extension.
        """
    )
    
    parser.add_argument('src_tmpl', type=str,
                       help='Source template string (e.g., "data/%%s.%%s")')
    parser.add_argument('dest_tmpl', type=str,
                       help='Destination template string (e.g., "output/det-%%s.%%s")')
    parser.add_argument('--mode_op', type=str, default='',
                       help='Operation mode, default: "" (normal)')
    parser.add_argument('--verbose', '-v', action='count', default=1,
                       help='Verbosity')
    parser.add_argument('--quiet', '-q', action='store_true', help='Minimal verbosity')

    args = parser.parse_args()
    if args.quiet: args.verbose = 0

    # Create mode dictionary
    mode = {'op': args.mode_op, 'verbose': args.verbose}

    # Read info file and convert to dict
    info_file = args.src_tmpl % ("info", "csv")
    reduce_info = pd.read_csv(info_file).iloc[0].to_dict()

    # Load CSV data and run the plotting function
    plot_data = cs.load_csv_files(args.src_tmpl, ['visit-time'])
    rv = plot_drm_visit_times(reduce_info, plot_data, args.dest_tmpl, mode)
    return rv


if __name__ == '__main__':
    rv = main()
    if rv is None:
        print(f"Plots failed. Error signaled.", file=sys.stderr)
    else:
        print(f"Done. Wrote {len(rv)} plot(s).")
    sys.exit(1 if rv is None else 0)
